### DISCUSSION TITLE
Can we add DiskANN in the Milvus size tool?

### DISCUSSION THREAD

**Discussion by pycui** [milvus-io/milvus#35894] [Q&A and General discussion]
It's not the index type list currently. 

Also, I'm wondering if the DiskANN index will load scalar data into memory. This matters for estimating the memory need for long string fields. Thanks

**Comment by yhmo** [comment]
We have engineers working on this task. The sizing tool is based on practice experience, which means they will spend lots of time to test.

**Reply by pycui** [reply]
Thanks, some initial estimate would be also helpful.

BTW, the sizing tool currently shows installment guide using old helm chart repo, which contains bugs that are fixed in latest, i.e. 
```

helm repo add milvus https://milvus-io.github.io/milvus-helm/
helm repo update
helm install my-release milvus/milvus -f helmConfigYml.yml
```
Would be good to fix it.

**Reply by LoveEachDay** [reply]
@pycui Good catch. We'll fix the values output from sizing tool shortly.

**Reply by pycui** [reply]
👍 
BTW, I'm wondering if @yhmo @LoveEachDay can help answer my question on DiskANN-indexed collection: does this kind of collection load all rows' scalar fields into memory? or only some rows' scalar fields? Thanks.

**Reply by yhmo** [reply]
If mmap is enabled: https://milvus.io/docs/mmap.md, scalar fields are mapped into memory. Otherwise, scalar fields are loaded into memory,

**Reply by pycui** [reply]
Documentation mentions "Memory mapping is not supported for DiskANN or GPU-class indexes.". Does this mean DiskANN-indexed collection always load _ALL_ rows' scalar fields into memory when collection is loaded?

**Reply by yhmo** [reply]
Two different things: load vector index and load scalar index.
Three kinds of vector indexes: in-memory index(IVF_FLAT/HNSW, etc), diskann index and gpu index
1. If mmap is enabled: 
- in-memory index is mapped in memory; 
- diskann index is directly loaded in memory(partly in memory, the majority is on disk); 
- gpu index is directly loaded into gpu memory; 
- scalar fields are mapped to memory

2. If mmap is disabled: 
- in-memory index is directly loaded in memory; 
- diskann index is directly loaded in memory(partly in memory, the majority is on disk); 
- gpu index is directly loaded into gpu memory; 
- scalar fields are directly loaded into memory.

**Reply by pycui** [reply]
I want to be a little bit more clear: what about rows? if DiskANN, do we still load _all_ rows' scalar fields into memory (either mapped or loaded)? This is very relevant for estimating the memory footprint, which isn't clear (e.g. Sizing tool doesn't include diskAnn and doesn't consider scalar fields). Hope you can shed some light

**Reply by yhmo** [reply]
Milvus data is stored column-based, not row-based. Each field is a column, while a row is a logic concept that treats the data line by line.
Since the storage is column-based, fields can be loaded separately.

Let's say a collection has three fields: ID(int64 type), Vector(float-vector type), Title(varchar type). "Title" is a typical scalar field.
Now we insert the data into the collection:
```
collection.insert(data={"ID": 101, "Vector": [1, 2, 3], "Title": "aaa"})    # this is one row
collection.insert(data={"ID": 102, "Vector": [2, 3, 4], "Title": "bbb"})    # this is one row
collection.insert(data={"ID": 103, "Vector": [3, 4, 5], "Title": "ccc"})    # this is one row
```

In the storage, [101, 102, 103] is stored as a column-base file. [[1, 2,3], [2,3,4], [3,4,5]] is stored as a column-based file. ["aaa", "bbb", "ccc"] is stored as a column-based file.

When we load the collection into memory, we can choose to load the file of "Vector", or load all files of all fields.

In data UI, you will see the data in a grid with 3 rows like this:
| ID | Vector | Title |
| :---: |:---:|:---:|
|101|[1,2,3]|"aaa"|
|102|[2,3,4]|"bbb"|
|103|[3,4,5]|"ccc"|

**Reply by pycui** [reply]
Thanks for the answer. This seems indicating to me that we shouldn't put rich scalar data in the collection, e.g. text data, even cut down to say 8K bytes, are still not suitable for large collection, because they will be either loaded or mapped to memory completely when they are loaded as part of the collection, meaning you either have high mem or high local storage. Using DiskANN-index won't help here, as the scalar fields are loaded separately. 

I understand we can put a reference (e.g. ID/path) in the scalar field, and fetch it separately from other storage, however I was hoping we can do a lot of processing together in Milvus, so it's a bit limiting. Thanks for the detailed reply though!

**Reply by yhmo** [reply]
DiskANN is a special index for vector field, not for scalar fields.

Index types for vector field:
- in-memory index: IVF_FLAT, IVF_SQ8, IVF_PQ, HNSW, SCANN  https://milvus.io/docs/index.md?tab=floating#In-memory-Index
- on-disk(partly) index: DISKANN  https://milvus.io/docs/disk_index.md#On-disk-Index
- GPU index: GPU_IVF_FLAT, GPU_IVF_PQ, GPU_CAGRA  https://milvus.io/docs/gpu_index.md#GPU-Index

Index types for scalar field: https://milvus.io/docs/index-scalar-fields.md
INVERTED, STL_SORT, TRIE

**Reply by pycui** [reply]
I know all these indexes. What I meant is there isn't a way to effectively limit the memory footprint of the collection (ideally you would expect Disk-based index to help here, but as you point, it doesn't help scalar fields). I have been testing 1 billion rows with 1024 dim vector + 8K byte scalar fields, and the current features all require a huge amount of memory to be used.

**Reply by yhmo** [reply]
Enable mmap can help reduce memory usage of scalar fields.
Currently, Milvus doesn't allow users to specify which scalar field to be loaded or unloaded, for some technical reasons. Personally, I think it might be a valuable feature in future versions.

**Reply by pycui** [reply]
in some k8s environment, mmap can also be undesired when ephemeral storage is limited. Ideally a disk-based index would just use the MinIO storage for a scalar field

**Reply by xiaofan-luan** [reply]
we do have a mode using minio bat due to the long latency of object storage it could take a minutes to finish the search
